modules/lib/global.d.ts

```diff
@@ -1,130 +1,163 @@
+     * log.warning('My JSON: %s', {a: 1});
+     *
+     * @example
+     * // Log JSON object using string
+     * log.error('My JSON: %s', JSON.stringify({a: 1}, null, 2));
+     *
+     * @global
+     * @namespace
      */
-    debug: (...args: unknown[]) => void;
+    declare const log: Log;
 
     /**
-     * Log info message.
+     * Javascript to Java bridge functions.
      *
-     * @param {Array} args... logging arguments.
+     * @example
+     * var bean = __.newBean('com.enonic.xp.MyJavaUtils');
+     *
+     * @example
+     * return __.toNativeObject(bean.findArray(arrayName));
+     *
+     * @global
+     * @namespace
      */
-    info: (...args: unknown[]) => void;
+    declare const __: DoubleUnderscore;
 
     /**
-     * Log warning message.
+     * This globally available function will load a JavaScript file and return the exports as objects.
+     * The function implements parts of the `CommonJS Modules Specification`.
      *
-     * @param {Array} args... logging arguments.
+     * @example
+     * // Require relative to this
+     * var other = require('./other.js');
+     *
+     * @example
+     * // Require absolute
+     * var other = require('/path/to/other.js');
+     *
+     * @example
+     * // Require without .js extension
+     * var other = require('./other');
+     *
+     * @param {string} path Path for javascript file (relative or absolute and .js ending is optional).
+     * @returns {object} Exports from loaded javascript.
+     * @global
      */
-    warning: (...args: unknown[]) => void;
+    declare const require: XpRequire;
 
     /**
-     * Log error message.
+     * Resolves a path to another file. Can use relative or absolute path.
      *
-     * @param {Array} args... logging arguments.
+     * @example
+     * // Resolve relative to this
+     * var path = resolve('./other.html');
+     *
+     * @example
+     * // Resolve absolute
+     * var path = resolve('/path/to/other.html');
+     *
+     * @param {string} path Path to resolve.
+     * @returns {*} Reference to an object.
+     * @global
      */
-    error: (...args: unknown[]) => void;
-};
-
-declare interface ScriptValue {
-    isArray(): boolean;
-
-    isObject(): boolean;
-
-    isValue(): boolean;
-
-    isFunction(): boolean;
+    declare const resolve: Resolve;
 
-    getValue(): unknown;
-
-    getKeys(): string[];
-
-    hasMember(key: string): boolean;
-
-    getMember(key: string): ScriptValue;
-
-    getArray(): ScriptValue[];
+}
 
-    getMap(): Record<string, unknown>;
+//──────────────────────────────────────────────────────────────────────────────
+// Exported types
+//──────────────────────────────────────────────────────────────────────────────
+export type NewBean = NewBeanDefinition;
+export type ScriptValue = ScriptValueDefinition;
 
-    getList(): object[];
+export interface App {
+    /**
+     * The name of the application.
+     *
+     * @type string
+     */
+    name: string;
```

Comment:
I have always preferred to have `app.name` be a string literal in my projects. This is the main reason I usually roll my own global-types.

```typescript

declare const app: {
  name: "no.item.www";
  version: string;
  config: Record<string, string | undefined>;
};
```

Because then I can do:
```typescript
type ArticleViewPart = PartComponent<`${typeof app.name}:article-view`>;

export function get() {
  const part = getComponent<ArticleViewPart>();
  ...
}
```

---

Could a solution be to not declare `app.name` in globals, but rely on interface merging with a file in the developers' code instead?

Reply:
I do see the use for and like your making app.name a string literal in projects.

I'm assuming this PR currently breaks merging, because
```const app``` is a global const, not an interface type.

I should read this:
https://www.typescriptlang.org/docs/handbook/declaration-merging.html

Let's say I make a
interface AppDefinition {...}

And
declare global {
 type App = AppDefinition
 declare const app: App;
}

export type App = AppDefinition

--

That should allow you to "interface merge" global.App?

Reply:
One cannot use declaration merging to change app.name from string to 'com.example'.

https://www.typescriptlang.org/docs/handbook/declaration-merging.html#merging-interfaces

Non-function members of the interfaces should be unique. If they are not unique, they must be of the same type. The compiler will issue an error if the interfaces both declare a non-function member of the same name, but of different types.

Reply:
A small addition to what Christian mentioned. Unfortunately, TypeScript does not allow changing the type of parameter during declaration merging. While it is technically possible to do this if `"skipLibCheck: true"` is enabled, the error won't disappear, and I cannot recommend this approach.

Sure, using a specific value as the type for `app.name` works well with custom global types. I do this myself in some projects. However, if the variable's value is used as a type, I don't see any reason why you shouldn't globally define `type AppName = 'my.app'` and use it instead of `typeof app.name`. I haven't encountered scenarios where we need to check the value of `app.name` and where we specifically need to know its exact type (exact string literal).

Reply:
I think keeping it the way it is is the easiest for new developers, so that is the way it should be! :)

It's better to widen the types if the model is easier to understand for new developers.

---

> I don't see any reason why you shouldn't globally define type `AppName = 'my.app'`

Very true! This is really a luxury I'm asking for. :)